scripts/testing/e2e-runner

```diff
@@ -0,0 +1,396 @@
+
+
+#
+# Derived configuration, set up after command line parsing
+#
+# RESULT_DIR: directory for publishing results of this test run
+# WORKTREE:   git worktree for this test run
+# LOG:        log file for this test run
+
+RESULT_DIR=""
+LOG=""
+WORKTREE=""
+
+usage() {
+    local exit_code="${exit_code:-0}"
+
+    echo "usage: ${1:-$SELF} [options] [<e2e-tests>]"
+    echo
+    echo "The possible options are:"
+    echo "  --source-proxies <file>  source proxy environment variables from file"
+    echo "  --remote|r <repo>        remote repository to clone/fetch from"
+    echo "  --branch|-b <branch>     name of the remote branch to test"
+    echo "  --local|-l <dir>         local directory to clone remote respository into"
+    echo "  --results|--out <dir>    top-level directory for publishing results"
+    echo "  --name|-n) <name>        subdirectory for results of this run"
+    echo "  --skip-worktree)         skip worktree preparation, assume already done"
+    echo "  --help)                  show this help"
+
+    exit $exit_code
+}
+
+info() {
+    echo 1>&2 "$@"
+}
+
+error() {
+    echo 1>&2 "error: $@"
+}
+
+fatal() {
+    echo 1>&2 "fatal error: $@"
+    exit 1
+}
+
+must-cd() {
+    if ! cd "$1"; then
+        fatal "failed to change directory to $1"
+    fi
+}
+
+must-mkdir() {
+    local dir="$1"; shift
+    if ! mkdir -p "$dir"; then
+        fatal "failed to create ${*:+$* }directory $dir"
+    fi
+}
+
+must-mv() {
+    if ! mv $*; then
+        fatal "mv $* failed"
+    fi
+}
+
+source_proxies() {
+    local var
+
+    if [ -z "$PROXIES_SRC" ]; then
+        info "no proxies to set up..."
+        return 0
+    fi
+
+    info "setting up proxies..."
+    if ! source "$PROXIES_SRC"; then
+        fatal "failed to source proxies from $src"
+    fi
+
+    for var in ${!http*} ${!HTTP*} ${!no_*} ${!NO_*}; do
+        case $var in
+            *_proxy|*_PROXY)
+                info "exporting proxy envrionment variable $var (with value ${!var})..."
+                export $var
+                ;;
+        esac
+    done
+}
+
+prepare_worktree() {
+    local worktree_dir="$(dirname $CLONED_REPO)/worktree"
+
+    WORKTREE="$worktree_dir/$RESULT_NAME"
+    case "$SKIP_WORKTREE" in
+        1|true|yes|y)
+            WORKTREE="$(realpath $WORKTREE)"
+            must-cd "$WORKTREE"
+            return 0
+            ;;
+    esac
+
+    if [ ! -e "$CLONED_REPO" ]; then
+        must-mkdir "$CLONED_REPO" "local repo"
+        must-cd "$CLONED_REPO"
+        if ! git clone "$REMOTE_REPO" .; then
+            fatal "failed to clone $REMOTE_REPO to $CLONED_REPO"
+        fi
+    else
+        must-cd "$CLONED_REPO"
+        if ! git remote -v | grep origin | grep -q "$REMOTE_REPO"; then
+            fatal "existing clone $CLONED_REPO has other origin than $REMOTE_REPO"
+        fi
+        if ! git fetch origin; then
+            fatal "failed to fetch $REMOTE_REPO"
+        fi
+    fi
+
+    must-mkdir "$worktree_dir" "worktree root"
```

Comment:
Is there a reason we're pre-mkdir'ing the directories (git clone and worktree) and not let git do it?

Reply:
The only reason is to have a better understanding and clearer error messages in case we don't have the necessary permissions to create the directories. But I'm fine with letting `git clone` and `git worktree add` fail with whatever messages they see fit, too.